File: kotlin/src/main/kotlin/aethernet/vaultshard/VaultShardRequestService.kt
// SPDX-License-Identifier: MIT

// Wire binding for aether-vault shard requests (Phase-2 extension). Binds
// PacketType.VaultShardRequest (42) to the mesh: ask peers for an erasure-coded
// shard by hash, and surface inbound requests via onShardRequested (the host answers
// from IVaultService if it holds the shard). Port of the C# reference
// (VaultShardRequestService).

package aethernet.vaultshard

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import aethernet.voice.JsonReader

/**
 * A request for a single erasure-coded shard the requester needs to recover a file.
 * Doubles as the canonical wire payload for [PacketType.VaultShardRequest] (42) and
 * the [VaultShardRequestService.onShardRequested] event arg. Field order: shard_hash,
 * requester_uhid. snake_case keys, no whitespace.
 *
 * Built by hand (no kotlinx.serialization — AOSP Soong forbids it) with a
 * StringBuilder so the emitted bytes match the C# reference exactly. Byte-identity
 * gate: fixtures/vaultshard/vectors.json.
 */
data class VaultShardRequest(
    val shardHash: String = "",
    val requesterUhid: String = "",
) {
    /** Serialize to the canonical UTF-8 JSON wire bytes. */
    fun toJsonBytes(): ByteArray {
        val sb = StringBuilder()
        sb.append('{')
        sb.append("\"shard_hash\":\"").append(jsonEscape(shardHash)).append("\",")
        sb.append("\"requester_uhid\":\"").append(jsonEscape(requesterUhid)).append('"')
        sb.append('}')
        return sb.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        /** Parse canonical wire bytes into a request, or null if malformed / shard_hash missing. */
        fun fromJson(json: String): VaultShardRequest? {
            val shardHash = JsonReader.readString(json, "shard_hash") ?: return null
            if (shardHash.isEmpty()) return null
            return VaultShardRequest(
                shardHash = shardHash,
                requesterUhid = JsonReader.readString(json, "requester_uhid") ?: "",
            )
        }

        private fun jsonEscape(s: String): String {
            val sb = StringBuilder()
            for (c in s) {
                when (c) {
                    '\\' -> sb.append("\\\\")
                    '"' -> sb.append("\\\"")
                    '\n' -> sb.append("\\n")
                    '\r' -> sb.append("\\r")
                    '\t' -> sb.append("\\t")
                    else -> sb.append(c)
                }
            }
            return sb.toString()
        }
    }
}

/**
 * Binds [PacketType.VaultShardRequest] (42) to the mesh: ask peers for a shard, and
 * surface inbound shard requests via [onShardRequested] (the host answers from
 * IVaultService if it holds the shard). Transport for the aether-vault erasure-coded-
 * storage extension. Mirrors the C# VaultShardRequestService.
 */
class VaultShardRequestService(private val sender: MeshSender) {

    /** Raised when a peer requests a shard. */
    var onShardRequested: ((VaultShardRequest) -> Unit)? = null

    /**
     * Broadcast a request for [shardHash] (requester = this node's localUhid). Returns
     * the number of peers reached.
     */
    suspend fun requestShard(shardHash: String): Int {
        require(shardHash.isNotEmpty()) { "shardHash must not be empty" }
        val payload = VaultShardRequest(shardHash = shardHash, requesterUhid = sender.localUhid)
        val packet = MeshPacket(
            type = PacketType.VaultShardRequest,
            sourceUhid = sender.localUhid,
            destinationUhid = "*",
            ttl = AetherNetConstants.DEFAULT_TTL,
            payload = payload.toJsonBytes(),
        )
        return sender.broadcast(packet)
    }

    /**
     * Process an inbound [PacketType.VaultShardRequest]. Returns false on wrong type or
     * malformed payload; on success fires [onShardRequested] and returns true.
     */
    suspend fun handle(packet: MeshPacket): Boolean {
        if (packet.type != PacketType.VaultShardRequest) return false
        val body = VaultShardRequest.fromJson(packet.payload.toString(Charsets.UTF_8)) ?: return false
        onShardRequested?.invoke(body)
        return true
    }
}

File: typescript/src/vaultshard/VaultShardRequestService.ts
/**
 * WIRE binding for the aether-vault erasure-coded-storage extension
 * (PacketType.VaultShardRequest = 42). A thin transport service: a node broadcasts a request for a
 * shard it needs to recover a file; inbound requests surface via onShardRequested (the host answers
 * from its IVaultService if it holds the shard).
 *
 * Mirrors the C# AetherNet.Vault.VaultShardRequestService.
 *
 * Wire payload (byte-identity gate — fixtures/vaultshard/vectors.json): UTF-8 JSON, snake_case keys,
 * field order shard_hash, requester_uhid — no whitespace.
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";

/**
 * A peer's request for an erasure-coded shard. Surfaced on the receiving node when a
 * VaultShardRequest arrives. `requesterUhid` is the UHID of the node that needs the shard.
 */
export interface VaultShardRequest {
  /** SHA-256 hex of the requested shard. */
  shardHash: string;
  /** UHID of the node that needs the shard. */
  requesterUhid: string;
}

/** JSON payload for a VaultShardRequest packet (same fields as the surfaced request). */
export type VaultShardRequestPayload = VaultShardRequest;

/**
 * Canonical VaultShardRequest payload serialization — MUST be byte-identical across all language
 * ports (fixtures/vaultshard/vectors.json): snake_case keys, field order shard_hash,
 * requester_uhid, no whitespace.
 */
export function serializeVaultShardRequestPayload(p: VaultShardRequestPayload): string {
  return JSON.stringify({
    shard_hash: p.shardHash,
    requester_uhid: p.requesterUhid,
  });
}

/** Parse a canonical VaultShardRequest payload back into camelCase fields. */
export function deserializeVaultShardRequestPayload(bytes: Uint8Array): VaultShardRequestPayload {
  const data = JSON.parse(new TextDecoder().decode(bytes)) as {
    shard_hash?: string;
    requester_uhid?: string;
  };
  return {
    shardHash: data.shard_hash ?? "",
    requesterUhid: data.requester_uhid ?? "",
  };
}

/**
 * Binds PacketType.VaultShardRequest (42) to the mesh. Transport for the aether-vault erasure-coded
 * storage extension: broadcast a request for a shard, and surface inbound shard requests via
 * onShardRequested.
 */
export class VaultShardRequestService {
  /** Raised when a peer requests a shard. */
  onShardRequested?: (request: VaultShardRequest) => void;

  constructor(private readonly sender: IMeshSender) {}

  /**
   * Broadcast a request for `shardHash` (dest "*", default TTL). The requester is this node's
   * localUhid. Returns the number of peers reached directly.
   */
  async requestShard(shardHash: string): Promise<number> {
    if (!shardHash) throw new Error("shardHash must not be empty");

    const body = new TextEncoder().encode(
      serializeVaultShardRequestPayload({ shardHash, requesterUhid: this.sender.localUhid }),
    );

    const packet = new MeshPacket();
    packet.type = PacketType.VaultShardRequest;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = "*";
    packet.ttl = DEFAULT_TTL;
    packet.payload = body;

    return this.sender.broadcast(packet);
  }

  /**
   * Process an incoming PacketType.VaultShardRequest packet: surface it via onShardRequested.
   * Returns false for the wrong packet type, a malformed payload, or an empty shard hash.
   */
  async handle(packet: MeshPacket): Promise<boolean> {
    if (packet.type !== PacketType.VaultShardRequest) return false;

    let body: VaultShardRequestPayload | undefined;
    try {
      body = deserializeVaultShardRequestPayload(packet.payload);
    } catch {
      return false;
    }
    if (!body || !body.shardHash) return false;

    this.onShardRequested?.(body);
    return true;
  }
}
